### PR TITLE
String#forcible_encoding?

### PR DESCRIPTION
**Pull request by kddnewton**
Introduce String#forcible_encoding?(encoding) to check if it is possible to cleanly force an encoding onto a string without it breaking.

### CODE REVIEW

File: string.c [hunk 1]
```diff
@@ -11310,6 +11310,51 @@ rb_str_force_encoding(VALUE str, VALUE enc)
...
+
+    rb_enc_associate_index(str, idx);
+
+    ENC_CODERANGE_CLEAR(str);
+    VALUE result = RBOOL(rb_enc_str_coderange(str) != ENC_CODERANGE_BROKEN);
+
+    rb_enc_associate_index(str, original_encoding);
+    ENC_CODERANGE_SET(str, original_coderange);
+
+    return result;
```
Comment (nobu) [inline string.c:11355]:
```suggestion
    if (ENC_CODERANGE(str) == ENC_CODERANGE_7BIT && encoding && rb_enc_asciicompat(encoding)) {
        return Qtrue;
    }

    return RBOOL(rb_enc_str_coderange_scan(str, encoding) != ENC_CODERANGE_BROKEN);
```

File: string.c [hunk 2]
```diff
@@ -11310,6 +11310,51 @@ rb_str_force_encoding(VALUE str, VALUE enc)
...
+ *    "\xc2\xa1".forcible_encoding?("UTF-8") # => true
+ *    "\xc2".forcible_encoding?("UTF-8")     # => false
+ *    "\x80".forcible_encoding?("UTF-8")     # => false
+ */
+
+static VALUE
+rb_str_forcible_encoding_p(VALUE str, VALUE enc)
+{
+    str_modifiable(str);
+
```
Comment (nobu) [inline string.c:11329]:
```suggestion
```

File: test/ruby/test_m17n.rb [hunk 3]
```diff
@@ -1579,6 +1579,17 @@ def test_force_encoding
...
 
+  def test_forcible_encoding
+    a = "\xC3"
+    refute(a.forcible_encoding?(Encoding::UTF_8))
+
+    a = "\xC3\xB6"
+    assert(a.forcible_encoding?(Encoding::UTF_8))
+
+    a = "\x80\x00"
+    assert(a.forcible_encoding?(Encoding::UTF_16LE))
```
Comment (nobu) [inline test/ruby/test_m17n.rb:1590]:
```suggestion
    a = -"\xC3"
    assert_not_operator(a, :forcible_encoding?, Encoding::UTF_8)

    a = -"\xC3\xB6"
    assert_operator(a, :forcible_encoding?, Encoding::UTF_8)

    a = -"\x80\x00"
    assert_operator(a, :forcible_encoding?, Encoding::UTF_16LE)
```

**Comment by nobu** [comment]
`"\xc2A".valid_encoding?` is false, but  should `"\xc2A".forcible_encoding?("UTF-8")` be true?

**Comment by kddnewton** [comment]
> `"\xc2A".valid_encoding?` is false, but should `"\xc2A".forcible_encoding?("UTF-8")` be true?

That's a good point! I think it should probably return `false`. I will update the PR.

**Pull request closed without merging**